api/pkg/auth/auth.go

```diff
@@ -0,0 +1,151 @@
+	"net/http"
+
+	"github.com/markbates/goth/gothic"
+	"github.com/tektoncd/hub/api/gen/log"
+	"github.com/tektoncd/hub/api/pkg/app"
+	"github.com/tektoncd/hub/api/pkg/db/model"
+	"gorm.io/gorm"
+)
+
+type service struct {
+	app.Service
+	api app.Config
+}
+
+type request struct {
+	db            *gorm.DB
+	log           *log.Logger
+	defaultScopes []string
+	jwtConfig     *app.JWTConfig
+}
+
+type ProviderList struct {
+	Provider      string
+	EnterpriseUrl string
+	ClientId      string
+}
+
+var (
+	UI_URL string
+)
+
+type Service interface {
+	AuthCallBack(res http.ResponseWriter, req *http.Request)
+	HubAuthenticate(res http.ResponseWriter, req *http.Request)
+}
+
+// New returns the resource service implementation.
+func New(api app.Config) Service {
+	return &service{
+		Service: api.Service("auth"),
+		api:     api,
+	}
+}
+
+func Authenticate(res http.ResponseWriter, req *http.Request) {
+	UI_URL = req.FormValue("redirect_uri")
+	gothic.BeginAuthHandler(res, req)
+}
+
+func (s *service) AuthCallBack(res http.ResponseWriter, req *http.Request) {
+	ghUser, err := gothic.CompleteUserAuth(res, req)
+	if err != nil {
+		fmt.Fprintln(res, err)
+		return
+	}
+
+	r := request{
+		db:            s.DB(context.Background()),
+		log:           s.Logger(context.Background()),
+		defaultScopes: s.api.Data().Default.Scopes,
+		jwtConfig:     s.api.JWTConfig(),
+	}
+
+	params := req.URL.Query()
+
+	if err = r.insertData(ghUser, params.Get("code")); err != nil {
+		s.Logger(context.Background()).Error(err)
+		res.Header().Set("Location", fmt.Sprintf("%s?status=%d", UI_URL, http.StatusBadRequest))
+		res.WriteHeader(http.StatusTemporaryRedirect)
+	}
+
+	res.Header().Set("Location", fmt.Sprintf("%s?status=%d&code=%s", UI_URL, http.StatusOK, params.Get("code")))
+	res.WriteHeader(http.StatusTemporaryRedirect)
+}
+
+func (s *service) HubAuthenticate(res http.ResponseWriter, req *http.Request) {
+
+	code := req.FormValue("code")
+
+	r := request{
+		db:            s.DB(context.Background()),
+		log:           s.Logger(context.Background()),
+		defaultScopes: s.api.Data().Default.Scopes,
+		jwtConfig:     s.api.JWTConfig(),
+	}
+
+	var User model.User
+	// Check if user exist
+	q := r.db.Model(&model.User{}).
+		Where("code = ?", code)
+
+	err := q.First(&User).Error
+	if err != nil {
+		r.log.Error(err)
+	}
+
+	if err := r.db.Model(&model.User{}).Where("github_login = ?", User.GithubLogin).Update("code", "").Error; err != nil {
+		r.log.Error(err)
+	}
+
+	// gets user scopes to add in jwt
+	scopes, err := r.userScopes(&User)
+	if err != nil {
+		// return nil, err
+		r.log.Error(err)
+	}
+
+	user, err := r.createTokens(&User, scopes)
+	if err != nil {
+		r.log.Error(err)
+	}
+
+	res.WriteHeader(http.StatusOK)
+	if err := json.NewEncoder(res).Encode(user); err != nil {
+		r.log.Error(err)
+	}
+}
+
+func List(res http.ResponseWriter, req *http.Request) {
+	type providerList []ProviderList
+	providers := providerList{
+		ProviderList{
+			Provider:      "github",
```

Comment:
could you update the response , I think Provider is needed for ui now and can we convert field `Provider` to `name` ?

Reply:
and also needs response in format of `{data: providers}`

Reply:
Yes will update it 👍🏻